tt_helpers.go

```diff
@@ -0,0 +1,121 @@
+package tupleconv
+
+import "errors"
+
+// ErrInvalidSpaceFmt is an error of the space format.
+var ErrInvalidSpaceFmt = errors.New("invalid space format")
+
+const (
+	kType       = "type"
+	kIsNullable = "is_nullable"
+)
+
+const (
+	defaultIgnoreChars   = ""
+	defaultDecSeparators = "."
+	defaultNullValue     = ""
+)
+
+// MappersTable can return a Mapper[string] based on the type name.
+type MappersTable interface {
+	getMapper(typ TypeName) (Mapper[string], bool)
+}
+
+// ParsersTTTable is basic tarantool parsers table.
+type ParsersTTTable struct{}
+
+// getMapper is the implementation of MappersTable for MappersTTTable.
+func (ptable *ParsersTTTable) getMapper(typ TypeName) (Mapper[string], bool) {
+	mapper, ok := parserByTTType[typ]
+	return mapper, ok
+}
+
+// parserByTTType is a mapping from tarantool typenames to parsers.
+var parserByTTType = map[TypeName]Mapper[string]{
+	TypeBoolean:   NewBoolParser(),
+	TypeString:    NewStringParser(),
+	TypeUnsigned:  NewUIntParser(defaultIgnoreChars),
+	TypeDatetime:  NewDateTimeParser(),
+	TypeUUID:      NewUUIDParser(),
+	TypeMap:       NewMapParser(),
+	TypeArray:     NewSliceParser(),
+	TypeVarbinary: NewBinaryParser(),
+
+	TypeDouble:  NewFloatParser(defaultIgnoreChars, defaultDecSeparators),
+	TypeDecimal: NewDecimalParser(defaultIgnoreChars, defaultDecSeparators),
+
+	TypeInteger: MakeSequenceMapper([]Mapper[string]{
+		NewUIntParser(defaultIgnoreChars),
+		NewIntParser(defaultIgnoreChars),
+	}),
+
+	TypeNumber: MakeSequenceMapper([]Mapper[string]{
+		NewUIntParser(defaultIgnoreChars),
+		NewIntParser(defaultIgnoreChars),
+		NewFloatParser(defaultIgnoreChars, defaultDecSeparators),
+	}),
+
+	TypeAny: MakeSequenceMapper([]Mapper[string]{
+		NewUIntParser(defaultIgnoreChars),
+		NewIntParser(defaultIgnoreChars),
+		NewFloatParser(defaultIgnoreChars, defaultDecSeparators),
+		NewDecimalParser(defaultIgnoreChars, defaultDecSeparators),
+		NewBoolParser(),
+		NewDateTimeParser(),
+		NewIntervalParser(),
+		NewUUIDParser(),
+		NewStringParser(), // Always success.
+	}),
+
+	// Same as any by default.
+	TypeScalar: MakeSequenceMapper([]Mapper[string]{
+		NewUIntParser(defaultIgnoreChars),
+		NewIntParser(defaultIgnoreChars),
+		NewFloatParser(defaultIgnoreChars, defaultDecSeparators),
+		NewDecimalParser(defaultIgnoreChars, defaultDecSeparators),
+		NewBoolParser(),
+		NewDateTimeParser(),
+		NewIntervalParser(),
+		NewUUIDParser(),
+		NewStringParser(), // Always success.
+	}),
+
+	TypeInterval: NewIntervalParser(),
+}
+
+var ttParsers MappersTable = &ParsersTTTable{}
+
+// NewParserFor returns parser by the type name and `isNullable` property.
+var NewParserFor = func(typ TypeName, isNullable bool) (Mapper[string], error) {
+	parser, isSupported := ttParsers.getMapper(typ)
+	if !isSupported {
+		return nil, NewErrUnexpectedType(typ)
+	}
+	if isNullable {
+		parser = MakeSequenceMapper([]Mapper[string]{
+			parser,
+			NewNullParser(defaultNullValue),
+		})
```

Comment:
How should work a `StringParser` + `NullParser`?

Reply:
https://github.com/tarantool/go-tupleconv/blob/askalt/no-gh-initial-logic/tt_helpers_test.go#L245-L247

Reply:
Where is?

Reply:
https://github.com/tarantool/go-tupleconv/blob/askalt/no-gh-initial-logic/tt_test.go#L275-L277

Reply:
In default implementation - no.

Reply:
@psergee @DifferentialOrange is that a problem?

Reply:
Changed, now, firstly trying to convert to null.
https://github.com/tarantool/go-tupleconv/blob/askalt/no-gh-initial-logic/tt.go#L201-L204

Reply:
> You can find such case [here](https://github.com/tarantool/go-tupleconv/blob/askalt/no-gh-initial-logic/tt_helpers_test.go#L245-L247).

Better add a permalink with commit next time -- there is no such file now.